### PR TITLE
Fix link failing when runtime built with -g

### LINKED ISSUE #957: Link fails when runtime built with -g
If the runtime directory is built with `make CFLAGS=-g`, then in the `test` directory `make 14-bug` gets the error:

```
llvm-link-12 -o result/floatsubtypecompare-v.bc ll/floatsubtypecompare-v.ll ../../../runtime/balrt_inline.bc
inlinable function call in a function with debug info must have a !dbg location
  %27 = call i8 addrspace(1)* @_bal_float_to_tagged(double 1.000000e+00)
inlinable function call in a function with debug info must have a !dbg location
  %29 = call i8 addrspace(1)* @_bal_float_to_tagged(double 2.000000e+00)
inlinable function call in a function with debug info must have a !dbg location
  %31 = call i8 addrspace(1)* @_bal_float_to_tagged(double 3.000000e+00)
llvm-link-12: error: linked module is broken!
```

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Fix link failing when runtime built with -g
Resolves #957

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by heshanpadmasiri** [comment]
@jclark while this will fix #957 there seems to be another existing (in current main branch) for some test cases when runtime is build with main. It causes llvm to segfault giving an error message as fallows in few test cases (such as [`02-misc/stackoverflow-p.bal`](https://github.com/ballerina-platform/nballerina/blob/main/compiler/testSuite/02-misc/stackoverflow-p.bal)):
```
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /usr/lib/llvm-12/bin/clang -cc1 -triple x86_64-pc-linux-gnu -emit-obj --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name 1-v.bc -mrelocation-model static -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -fno-split-dwarf-inlining -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -resource-dir /usr/lib/llvm-12/lib/clang/12.0.1 -O2 -fdebug-compilation-dir /home/heshan/work/nballerina/test/out/04-typecast -ferror-limit 19 -fgnuc-version=4.2.1 -vectorize-loops -vectorize-slp -faddrsig -o /tmp/1-v-b8a31e.o -x ir result/1-v.bc
1.	Code generation
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /usr/lib/llvm-12/bin/clang -cc1 -triple x86_64-pc-linux-gnu -emit-obj --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name 2-v.bc -mrelocation-model static -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -fno-split-dwarf-inlining -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -resource-dir /usr/lib/llvm-12/lib/clang/12.0.1 -O2 -fdebug-compilation-dir /home/heshan/work/nballerina/test/out/04-typecast -ferror-limit 19 -fgnuc-version=4.2.1 -vectorize-loops -vectorize-slp -faddrsig -o /tmp/2-v-e8462f.o -x ir result/2-v.bc
1.	Code generation
 #0 0x00007febb6c76871 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd26871)
 #1 0x00007febb6c749a0 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd249a0)
 #2 0x00007febb6c76f82 (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd26f82)
 #3 0x00007febbe7c13c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x143c0)
 #4 0x00007febb7412700 llvm::DIE::getUnitDie() const (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14c2700)
 #5 0x00007febb742646f llvm::DwarfDebug::finalizeModuleInfo() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14d646f)
 #6 0x00007febb7426b38 llvm::DwarfDebug::endModule() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14d6b38)
 #7 0x00007febb73fc191 llvm::AsmPrinter::doFinalization(llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14ac191)
 #8 0x00007febb6db0161 llvm::FPPassManager::doFinalization(llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xe60161)
 #9 0x00007febb6daabf2 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xe5abf2)
#10 0x00007febbd014ac6 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x15a9ac6)
#11 0x00007febbd2beffc clang::CodeGenAction::ExecuteAction() (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1853ffc)
#12 0x00007febbd9f8456 clang::FrontendAction::Execute() (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1f8d456)
#13 0x00007febbd980331 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1f15331)
#14 0x00007febbda5f8e6 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1ff48e6)
#15 0x0000000000412ab9 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/lib/llvm-12/bin/clang+0x412ab9)
#16 0x0000000000410dbf (/usr/lib/llvm-12/bin/clang+0x410dbf)
#17 0x0000000000410bb6 main (/usr/lib/llvm-12/bin/clang+0x410bb6)
#18 0x00007febb5a340b3 __libc_start_main /build/glibc-sMfBJT/glibc-2.31/csu/../csu/libc-start.c:342:3
#19 0x000000000040ddce _start (/usr/lib/llvm-12/bin/clang+0x40ddce)
 #0 0x00007f132ae59871 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd26871)
 #1 0x00007f132ae579a0 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd249a0)
 #2 0x00007f132ae59f82 (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd26f82)
 #3 0x00007f13329a43c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x143c0)
 #4 0x00007f132b5f5700 llvm::DIE::getUnitDie() const (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14c2700)
 #5 0x00007f132b60946f llvm::DwarfDebug::finalizeModuleInfo() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14d646f)
 #6 0x00007f132b609b38 llvm::DwarfDebug::endModule() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14d6b38)
 #7 0x00007f132b5df191 llvm::AsmPrinter::doFinalization(llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0x14ac191)
 #8 0x00007f132af93161 llvm::FPPassManager::doFinalization(llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xe60161)
 #9 0x00007f132af8dbf2 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xe5abf2)
#10 0x00007f13311f7ac6 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x15a9ac6)
#11 0x00007f13314a1ffc clang::CodeGenAction::ExecuteAction() (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1853ffc)
#12 0x00007f1331bdb456 clang::FrontendAction::Execute() (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1f8d456)
#13 0x00007f1331b63331 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1f15331)
#14 0x00007f1331c428e6 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/lib/x86_64-linux-gnu/libclang-cpp.so.12+0x1ff48e6)
#15 0x0000000000412ab9 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/lib/llvm-12/bin/clang+0x412ab9)
#16 0x0000000000410dbf (/usr/lib/llvm-12/bin/clang+0x410dbf)
#17 0x0000000000410bb6 main (/usr/lib/llvm-12/bin/clang+0x410bb6)
#18 0x00007f1329c170b3 __libc_start_main /build/glibc-sMfBJT/glibc-2.31/csu/../csu/libc-start.c:342:3
#19 0x000000000040ddce _start (/usr/lib/llvm-12/bin/clang+0x40ddce)
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Ubuntu clang version 12.0.1-++20211029101322+fed41342a82f-1~exp1~20211029221816.4
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Ubuntu clang version 12.0.1-++20211029101322+fed41342a82f-1~exp1~20211029221816.4
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
This fix will also change the output in cases where we get a runtime panic if runtime is compiled with -g
ex:
from 
```
  panic: arithmetic overflow
  div1-p.bal:10
```
to
```
panic: arithmetic overflow
  _bal_error_construct /home/heshan/work/nballerina/runtime/error.c:10
  _bal_panic_construct /home/heshan/work/nballerina/runtime/panic.c:67
  div ../../../compiler/testSuite/01-int/div1-p.bal
  main (root) ../../../compiler/testSuite/01-int/div1-p.bal:6
  main /home/heshan/work/nballerina/runtime/main.c:11
```

File: compiler/modules/nback/typeTest.bal [hunk 1]
```diff
@@ -131,18 +131,21 @@ function buildNarrowRepr(llvm:Builder builder, Scaffold scaffold, Repr sourceRep
...
 
-function buildExactify(llvm:Builder builder, Scaffold scaffold, llvm:PointerValue tagged, t:SemType targetType) returns llvm:PointerValue {
+function buildExactify(llvm:Builder builder, Scaffold scaffold, llvm:PointerValue tagged, t:SemType targetType, bir:Position pos) returns llvm:PointerValue {
     t:Context tc = scaffold.typeContext();
     if t:mappingAtomicTypeRw(tc, targetType) == () && t:listAtomicTypeRw(tc, targetType) == () {
         return tagged;
     }
-    return <llvm:PointerValue>builder.call(scaffold.getRuntimeFunctionDecl(structureExactifyFunction),
-                                           [tagged, scaffold.getExactify(t:diff(targetType, t:READONLY))]);
+    scaffold.setDebugLocation(builder, pos, DEBUG_ORIGIN_CALL);
```
Comment (jclark) [inline compiler/modules/nback/typeTest.bal]: This pattern is being repeated all over the place.

Create a method on scaffold to do this.

**Review by jclark — COMMENTED** [review]
This all seems to be making things more difficult than it need to be.  Instructions only have one location.  We already call setDebugLocation once for each insn.  So it would be much simpler just to save the position there, and then subsequently make use of the saved location when generating code for this instruction.  We need to redesign how the setDebugLocation/clearDebugLocation functions work to handle this.  Can you propose a design to make this work?

**Comment by heshanpadmasiri** [comment]
> This all seems to be making things more difficult than it need to be. Instructions only have one location. We already call setDebugLocation once for each insn. So it would be much simpler just to save the position there, and then subsequently make use of the saved location when generating code for this instruction. We need to redesign how the setDebugLocation/clearDebugLocation functions work to handle this. Can you propose a design to make this work?

We can have calling setDebug location [here](https://github.com/ballerina-platform/nballerina/blob/16c4dee3660f681b38cd8283228b77c464574e07/compiler/modules/nback/basic.bal#L50) (We can detect this using the origin being other) setting a special field in builder (currently this call is ignored if we are not doing debug full) say instructionDbLoc (for this I'll add a function to builder as `public function setCurrentInstructionDebugLocation(Metadata? dbLocation)`). Then when [`call`](https://github.com/ballerina-platform/nballerina/blob/16c4dee3660f681b38cd8283228b77c464574e07/compiler/modules/print.llvm/llvm.bal#L1185) is called we can use that location automatically (between instruction we clear this).

File: compiler/modules/nback/typeTest.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/typeTest.bal]: fixed

File: compiler/modules/nback/scaffold.bal [hunk 2]
```diff
@@ -239,6 +240,17 @@ class Scaffold {
         }
     }
 
+    function buildRuntimeFunctionCall(llvm:Builder builder, RuntimeFunction rf, llvm:Value[] args) returns llvm:Value? {
+        return self.buildCall(builder, self.getRuntimeFunctionDecl(rf), args);
+    }
+
+    function buildCall(llvm:Builder builder, llvm:Function fn, llvm:Value[] args) returns llvm:Value? {
```
Comment (jclark) [inline compiler/modules/nback/scaffold.bal]: Don't make this a method.

File: compiler/modules/nback/scaffold.bal [hunk 3]
```diff
@@ -239,6 +240,17 @@ class Scaffold {
         }
     }
 
+    function buildRuntimeFunctionCall(llvm:Builder builder, RuntimeFunction rf, llvm:Value[] args) returns llvm:Value? {
```
Comment (jclark) [inline compiler/modules/nback/scaffold.bal]: Don't make this a method

File: compiler/modules/nback/scaffold.bal [hunk 4]
```diff
@@ -226,7 +227,7 @@ class Scaffold {
         return self.mod.llMod.getIntrinsicDeclaration(name);
     }
 
-    function getRuntimeFunctionDecl(RuntimeFunction rf) returns llvm:FunctionDecl {
+    private function getRuntimeFunctionDecl(RuntimeFunction rf) returns llvm:FunctionDecl {
```
Comment (jclark) [inline compiler/modules/nback/scaffold.bal]: Don't make this private

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/nback/scaffold.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/scaffold.bal]: fixed (I assume this is to move the function out of scaffold)

File: compiler/modules/nback/scaffold.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/scaffold.bal]: fixed (I assume this is to move the function out of scaffold)

File: compiler/modules/nback/scaffold.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/scaffold.bal]: fixed

**Comment by jclark** [comment]
We can also use the stored position for the *Err methods in scaffold.

**Comment by heshanpadmasiri** [comment]
> We can also use the stored position for the *Err methods in scaffold.

fixed

**Comment by heshanpadmasiri** [comment]
@jclark I changed the debug location creation to be lazy as we discussed with one change. We still need the `DEBUG_USAGE_OTHER` to represent cases where we need to set a location only if we are using fullDebug.

**Pull request merged**